File: javaproject/app/src/main/java/com/marcosvbras/robomarket/login/viewmodel/LoginViewModel.java
package com.marcosvbras.robomarket.login.viewmodel;

import android.annotation.SuppressLint;
import android.databinding.ObservableField;
import android.text.TextUtils;

import com.marcosvbras.robomarket.R;
import com.marcosvbras.robomarket.app.App;
import com.marcosvbras.robomarket.business.model.UserModel;
import com.marcosvbras.robomarket.utils.ErrorObservable;
import com.marcosvbras.robomarket.viewmodels.BaseViewModel;
import com.marcosvbras.robomarket.home.ui.activity.HomeActivity;
import com.marcosvbras.robomarket.register.ui.RegisterActivity;
import com.marcosvbras.robomarket.interfaces.BaseActivityCallback;

import io.reactivex.disposables.Disposable;

public class LoginViewModel extends BaseViewModel {

    private Disposable disposable;
    private UserModel userModel;
    public ObservableField<String> username = new ObservableField<>();
    public ObservableField<String> password = new ObservableField<>();
    public ErrorObservable usernameFieldError = new ErrorObservable();
    public ErrorObservable passwordFieldError = new ErrorObservable();

    public LoginViewModel(BaseActivityCallback activityCallback) {
        this.activityCallback = activityCallback;
        userModel = new UserModel();
    }

    @SuppressLint("CheckResult")
    public void signIn() {
        if(isFormValid()) {
            cleanupSubscriptions();

            userModel.login(username.get(), password.get())
                    .subscribe(next -> {
                        App.getInstance().setUser(next);
                        App.getInstance().writeUserCredentials(next.getObjectId(), next.getSessionToken());
                    }, error -> {
                        isLoading.set(false);
                        cleanupSubscriptions();
                        activityCallback.showDialogMessage(error.getMessage());
                    }, () -> {
                        isLoading.set(false);
                        cleanupSubscriptions();
                        activityCallback.openActivity(HomeActivity.class, true);
                    }, d -> {
                        isLoading.set(true);
                        disposable = d;
                    });
        }
    }

    public void register() {
//        activityCallback.openActivityForResult(RegisterActivity.class, null, Constants.Other.FINISH_LOGIN_ACTIVITY);
        activityCallback.openActivity(RegisterActivity.class, false);
    }

    private boolean isFormValid() {
        if(TextUtils.isEmpty(username.get())) {
            usernameFieldError.set(R.string.required_field);
            return false;
        }

        if(TextUtils.isEmpty(password.get())) {
            passwordFieldError.set(R.string.required_field);
            return false;
        }

        return true;
    }

    @Override
    public void cleanupSubscriptions() {
        if(disposable != null && !disposable.isDisposed())
            disposable.dispose();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        cleanupSubscriptions();
    }
}

File: javaproject/app/src/main/java/com/marcosvbras/robomarket/flows/login/viewmodel/LoginViewModel.kt
package com.marcosvbras.robomarket.flows.login.viewmodel

import android.databinding.ObservableField
import android.text.TextUtils
import com.marcosvbras.robomarket.R
import com.marcosvbras.robomarket.app.App
import com.marcosvbras.robomarket.business.model.UserModel
import com.marcosvbras.robomarket.flows.home.ui.activity.HomeActivity
import com.marcosvbras.robomarket.flows.register.ui.RegisterActivity
import com.marcosvbras.robomarket.interfaces.BaseActivityCallback
import com.marcosvbras.robomarket.utils.ErrorObservable
import com.marcosvbras.robomarket.viewmodels.BaseViewModel
import io.reactivex.disposables.Disposable

class LoginViewModel(private val callback: BaseActivityCallback) : BaseViewModel() {

    private var disposable: Disposable? = null
    private val userModel: UserModel = UserModel()
    var username = ObservableField<String>()
    var password = ObservableField<String>()
    var usernameFieldError = ErrorObservable()
    var passwordFieldError = ErrorObservable()

    private val isFormValid: Boolean
        get() {
            if (TextUtils.isEmpty(username.get())) {
                usernameFieldError.errorResource = R.string.required_field
                return false
            }

            if (TextUtils.isEmpty(password.get())) {
                passwordFieldError.errorResource = R.string.required_field
                return false
            }

            return true
        }

    fun signIn() {
        if (isFormValid) {
            cleanupSubscriptions()

            userModel.login(username.get()!!, password.get()!!)!!
                    .subscribe({ next ->
                        App.getInstance().user = next
                        App.getInstance().writeUserCredentials(next.objectId, next.sessionToken)
                    }, { error ->
                        cleanupSubscriptions()
                        callback.showDialogMessage(error.message!!)
                    }, {
                        cleanupSubscriptions()
                        callback.openActivity(HomeActivity::class.java, null,true)
                    }, { d ->
                        isLoading.set(true)
                        disposable = d
                    })
        }
    }

    fun register() {
        // callback.openActivityForResult(RegisterActivity.class, null, Constants.Other.FINISH_LOGIN_ACTIVITY);
        callback.openActivity(RegisterActivity::class.java, null,false)
    }

    override fun cleanupSubscriptions() {
        isLoading.set(false)
        disposable?.dispose()
    }

    override fun onCleared() {
        cleanupSubscriptions()
        super.onCleared()
    }
}